### DISCUSSION TITLE
Why does the GitHub Action use the "limited API" implicitly?

### DISCUSSION THREAD

**Discussion by bluenote10** [messense/maturin-action#274] [Q&A] [answered]
I'm using the a more or less default `CI.yml` generated by `maturin generate-ci github`.

My code makes usage of `pyo3::types::PyFunction` which is behind the a `not(Py_LIMITED_API)` compilation flag. My dependency towards pyo3 is specified via `pyo3 = "0.21.1"`, i.e., it is not using `features = ["abi3"]` to opt-in to the limited API. When I run `maturin` locally, it compiles fine.

However, on CI using the default `CI.yml` [I'm getting](https://github.com/megaplot/picoapp/actions/runs/9596789723/job/26464417948) compile errors from the `pyo3::types::PyFunction` usages.

```
error[E0432]: unresolved import `pyo3::types::PyFunction`
  --> src/conversion.rs:5:5
   |
5  | use pyo3::types::PyFunction;
   |     ^^^^^^^^^^^^^----------
   |     |            |
   |     |            help: a similar name exists in the module: `PyCFunction`
   |     no `PyFunction` in `types`
```

I can't figure out why the GitHub Action is using the limited API. My understanding from the documentation was that it should only do that if pyo3 dependency is used with `features = ["abi3"]`.

Interestingly, the build in the CI also seems to partially work, because the log seems to already have build some things successfully up to that point. I.e., above in the log there is:

<details>
<summary>Successful logs summary</summary>

```
📦 Built wheel for CPython 3.8 to dist/picoapp-0.1.0-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
   Compiling pyo3-build-config v0.21.2
   Compiling pyo3-macros-backend v0.21.2
   Compiling pyo3-ffi v0.21.2
   Compiling pyo3 v0.21.2
   Compiling pyo3-macros v0.21.2
   Compiling picoapp v0.1.0 (/home/runner/work/picoapp/picoapp)
    Finished `release` profile [optimized] target(s) in 14.71s
📦 Built wheel for CPython 3.9 to dist/picoapp-0.1.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
   Compiling pyo3-build-config v0.21.2
   Compiling pyo3-macros-backend v0.21.2
   Compiling pyo3-ffi v0.21.2
   Compiling pyo3 v0.21.2
   Compiling pyo3-macros v0.21.2
   Compiling picoapp v0.1.0 (/home/runner/work/picoapp/picoapp)
    Finished `release` profile [optimized] target(s) in 15.00s
📦 Built wheel for CPython 3.10 to dist/picoapp-0.1.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
   Compiling pyo3-build-config v0.21.2
   Compiling pyo3-macros-backend v0.21.2
   Compiling pyo3-ffi v0.21.2
   Compiling pyo3 v0.21.2
   Compiling pyo3-macros v0.21.2
   Compiling picoapp v0.1.0 (/home/runner/work/picoapp/picoapp)
    Finished `release` profile [optimized] target(s) in 14.97s
📦 Built wheel for CPython 3.11 to dist/picoapp-0.1.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
   Compiling pyo3-build-config v0.21.2
   Compiling pyo3-macros-backend v0.21.2
   Compiling pyo3-ffi v0.21.2
   Compiling pyo3 v0.21.2
   Compiling pyo3-macros v0.21.2
   Compiling picoapp v0.1.0 (/home/runner/work/picoapp/picoapp)
    Finished `release` profile [optimized] target(s) in 15.00s
📦 Built wheel for CPython 3.12 to dist/picoapp-0.1.0-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
   Compiling pyo3-build-config v0.21.2
   Compiling pyo3-macros-backend v0.21.2
   Compiling pyo3-ffi v0.21.2
   Compiling pyo3 v0.21.2
   Compiling pyo3-macros v0.21.2
   Compiling picoapp v0.1.0 (/home/runner/work/picoapp/picoapp)
```

</details>

It is unclear to me what the GitHub action is executing there exactly when the error happens:

```
💥 maturin failed
  Caused by: Failed to build a native library through cargo
  Caused by: Cargo build finished with "exit status: 101": `env -u CARGO PYO3_ENVIRONMENT_SIGNATURE="pypy-3.8-64bit" PYO3_PYTHON="/usr/local/bin/pypy3.8" PYTHON_SYS_EXECUTABLE="/usr/local/bin/pypy3.8" "cargo" "rustc" "--features" "pyo3/extension-module" "--target" "x86_64-unknown-linux-gnu" "--message-format" "json-render-diagnostics" "--manifest-path" "/home/runner/work/picoapp/picoapp/Cargo.toml" "--release" "--lib"`
```

**Comment by messense** [comment, marked as answer]
`PyFunction` is not available on PyPy: https://docs.rs/pyo3/latest/pyo3/types/struct.PyFunction.html

**Reply by bluenote10** [reply]
Indeed, thanks for the hint!

To add some context: I was simply on the wrong track. `PyFunction` has a conditional compilation from both `Py_LIMITED_API` and `PyPy`. The build actually wasn't using the limited API, and the symbol was rather missing when compiled against PyPy.

I've now followed the advice from https://github.com/PyO3/maturin-action/discussions/209 and switched from `--find-interpreters` (which picks up various `pypy` interpreters from the environment) to using an explicit `--interpreters ...` list to limit it to CPython.
